nni/algorithms/compression/pytorch/quantization/quantizers.py

```diff
@@ -120,6 +123,222 @@ def quant_backward(tensor, grad_output, quant_type, scale, zero_point, qmin, qma
+        observer = self.all_observers[name][quant_type]
+        scale, zero_point = observer.calculate_qparams()
+        return scale, zero_point
+
+    def _quantize(self, x, scale, zero_point, qmin, qmax):
+        x = x / scale + zero_point
+        x = torch.clamp(x, qmin, qmax)
+        x = torch.round(x)
+        x = (x - zero_point) * scale
+        return x
+
+    def quantize_input(self, *inputs, wrapper, **kwargs):
+        if self.compressed:
+            module = wrapper.module
+            new_input = self._quantize(inputs[0],
+                                      module.input_scale,
+                                      module.input_zero_point,
+                                      module.input_qmin,
+                                      module.input_qmax)
+            list_inp = list(inputs)
+            list_inp[0] = new_input
+            inputs = tuple(list_inp)
+        else:
+            self.record(wrapper, 'input', inputs)
+        return inputs
+
+    def quantize_weight(self, wrapper, **kwargs):
+        module = wrapper.module
+        old_weight = module.weight
+        if self.compressed:
+            new_weight = self._quantize(old_weight,
+                                       module.weight_scale,
+                                       module.weight_zero_point,
+                                       module.weight_qmin,
+                                       module.weight_qmax)
+        else:
+            self.record(wrapper, 'weight', old_weight)
+            new_weight = old_weight
+        return new_weight
+
+    def quantize_output(self, output, wrapper, **kwargs):
+        if self.compressed:
+            module = wrapper.module
+            new_output = self._quantize(output,
+                                       module.output_scale,
+                                       module.output_zero_point,
+                                       module.output_qmin,
+                                       module.output_qmax)
+        else:
+            self.record(wrapper, 'output', output)
+            new_output = output
+        return new_output
+
+    def compress(self):
+        """
+        Calculate quantization information of each tensor. Note that the inference of
+        the compressed model will no longer update the corresponding. Instead, the quantization
+        process will be simulated, which is used to test the accuracy of the quantization.
+        """
+        modules_to_compress = self.get_modules_to_compress()
+        for layer, config in modules_to_compress:
+            module = layer.module
+            if "weight" in config.get("quant_types", []):
+                scale, zero_point = self.calculate_qparams(layer.name, 'weight')
+                module.register_buffer('weight_scale', scale.to(self.device))
+                module.register_buffer('weight_zero_point', zero_point.to(self.device))
+                # todo: recover old_weight to weight, because the compressed
```

Comment:
could you explain more about the case that should recover old_weight to weight?

Reply:
Because after being wrapped by `QuantizerModuleWrapper`, two things may happen to the weight of each layer:
1. The type may changed from `torch.nn.Parameter` to `torch.Tensor`
2. The weight of BN may have been folded.

Basically, we need to ensure that the structure/parameter types of the model are consistent before and after PTQ. In theory, we can use the original model to perform downstream tasks (like qat or deployment), since the current ptq will not change the weight of the model.  But users may also use the model exported by the ptq for the downstream tasks, so I think it is better to recover them.

Reply:
I have added some code logics in `export_model` for recovering weight

Reply:
1) suppose quantizer's export should export the model after quantization. it is not proper to export original model.

2) I notice the inconsistency among the supported quantizers. some quantizers deal with fold bn, some others not. some quantizers recover weight, some others not. could you explain the current status a little bit more?

Reply:
As discussed offline, I replace the origin weight with the pseudo-quantized weight.